src/mon/MonClient.cc

```diff
@@ -687,23 +688,49 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
 
 void MonClient::_add_conns(uint64_t global_id)
 {
-  uint16_t min_priority = std::numeric_limits<uint16_t>::max();
+  map<uint16_t, vector<unsigned>> rank_by_priority;
   for (const auto& m : monmap.mon_info) {
-    if (m.second.priority < min_priority) {
-      min_priority = m.second.priority;
-    }
+    rank_by_priority[m.second.priority].push_back(monmap.get_rank(m.first));
+  }
+  vector<unsigned> candidates;
+  if (!rank_by_priority.empty()) {
+    // pick mon by priority, lowest first
+    // see http://tracker.ceph.com/issues/5249 for the use case of priority
+    candidates = rank_by_priority.begin()->second;
   }
   vector<unsigned> ranks;
-  for (const auto& m : monmap.mon_info) {
-    if (m.second.priority == min_priority) {
-      ranks.push_back(monmap.get_rank(m.first));
+  if (candidates.size() > 1) {
```

Comment:
what if there is only a single element in `candidates` ?

Reply:
Good catch. All fixed.